multi-node/aws/cmd/kube-aws/command_render.go

```diff
@@ -18,30 +26,321 @@ var (
 )
 
 func init() {
+	kubeconfigTemplate = template.Must(template.New("kubeconfig").Parse(kubeconfigTemplateContents))
+
 	cmdRoot.AddCommand(cmdRender)
-	cmdRender.Flags().StringVar(&renderOpts.Output, "output", "", "Write output to file instead of stdout")
-	cmdRender.Flags().StringVar(&renderOpts.ParameterDefaultArtifactURL, "parameter-default-artifact-url", cluster.DefaultArtifactURL(VERSION), "Set the default location of kube-aws deployment artifacts in the rendered template")
+	cmdRender.Flags().StringVar(&renderOpts.ConfigPath, "config", "./cluster.yaml", "Path to config yaml file")
 }
 
 func runCmdRender(cmd *cobra.Command, args []string) {
-	tmpl, err := cluster.StackTemplateBody(renderOpts.ParameterDefaultArtifactURL)
+	cfg := cluster.NewDefaultConfig(VERSION)
+	err := cluster.DecodeConfigFromFile(cfg, renderOpts.ConfigPath)
+	if err != nil {
+		stderr("Unable to load cluster config: %v", err)
+		os.Exit(1)
+	}
+
+	if rootOpts.AssetDir == "" {
+		stderr("--asset-dir option is not specified")
+		os.Exit(1)
+	}
+	if err := initAssetDirectory(cfg); err != nil {
+		stderr("Error initializing asset directory: %v", err)
+		os.Exit(1)
+	}
+
+	tmpl, err := cluster.StackTemplateBody()
 	if err != nil {
 		stderr("Failed to generate template: %v", err)
 		os.Exit(1)
 	}
 
-	if renderOpts.Output == "" {
-		os.Stdout.WriteString(tmpl)
-	} else {
-		if err := ioutil.WriteFile(renderOpts.Output, []byte(tmpl), 0600); err != nil {
-			stderr("Failed writing output to %s: %v", renderOpts.Output, err)
-			os.Exit(1)
+	templatePath := filepath.Join(rootOpts.AssetDir, "template.json")
+	if err := ioutil.WriteFile(templatePath, []byte(tmpl), 0600); err != nil {
+		stderr("Failed writing output to %s: %v", templatePath, err)
+		os.Exit(1)
+	}
+}
+
+func initAssetDirectory(cfg *cluster.Config) error {
+	if _, err := os.Stat(rootOpts.AssetDir); err == nil {
+		return fmt.Errorf("Asset directory %s already exists!", rootOpts.AssetDir)
+	} else if !os.IsNotExist(err) {
+		return fmt.Errorf("Error stat-ing asset directory path %s: %v", rootOpts.AssetDir, err)
+	}
+
+	if err := copyDir(rootOpts.AssetDir, "./artifacts"); err != nil {
+		return fmt.Errorf("Error copying artifacts: %v", err)
+	}
+
+	inCfg, err := os.Open(renderOpts.ConfigPath)
+	if err != nil {
+		return err
+	}
+	defer inCfg.Close()
+
+	outCfgPath := filepath.Join(rootOpts.AssetDir, "cluster.yaml")
+	outCfg, err := os.OpenFile(outCfgPath, os.O_WRONLY|os.O_CREATE, 0700)
+	if err != nil {
+		return err
+	}
+	defer outCfg.Close()
+
+	if _, err := io.Copy(outCfg, inCfg); err != nil {
+		return fmt.Errorf("Error copying config: %v", err)
+	}
+
+	credentialsDir := filepath.Join(rootOpts.AssetDir, "credentials")
+	if err := os.Mkdir(credentialsDir, 0700); err != nil {
+		return fmt.Errorf("Error creating credentials directory %s: %v", credentialsDir, err)
+	}
+
+	if err := initTLS(cfg, credentialsDir); err != nil {
+		stderr("Failed initializing TLS infrastructure: %v", err)
+		os.Exit(1)
```

Comment:
Nit: this is inconsistent. The caller will os.Exit anyway when an error is returned.